promql/engine.go

```diff
@@ -706,117 +654,354 @@ func (ev *evaluator) eval(expr Expr) Value {
 	if err := contextDone(ev.ctx, "expression evaluation"); err != nil {
 		ev.error(err)
 	}
+	numSteps := int((ev.endTimestamp-ev.timestamp)/ev.interval) + 1
+
+	// rangeWrapper evaluates the given expressions as matrixes, and then for
+	// each step calls the given function with the expressions for that step.
+	// The return value is the combination of all the function call results.
+	rangeWrapper := func(f func([]Value, *evalNodeHelper) Vector, exprs ...Expr) Matrix {
```

Comment:
I find the documentation somewhat difficult to parse. I would:

1. drop the "as matrixes" (it's obvious from the return type);
2. use a less generic name than `f` for the function (e.g. `instantEval`) so you don't have to refer to it as "the provided function";
3. replace " the expressions for that step" with "the values computed for each expression at that step";
4. be more specific about how the results are combined (e.g. "is the combination into time series of ...".

Reply:
Changed, though I think `f` is fine as a function name argument.

Reply:
The reason I was suggesting a name change was purely for documentation purposes (i.e. the same reason why you don't necessarily call most integer arguments `i`).

But I have no strong feelings about it, code already looks a lot more readable. Thank you for doing the work, BTW.

Reply:
Thanks for your comments, they have been useful.